File: api/src/main/java/com/openlattice/edm/collection/CollectionTemplates.java
package com.openlattice.edm.collection;

import com.google.common.collect.Maps;

import java.util.Map;
import java.util.Objects;
import java.util.UUID;

public class CollectionTemplates {

    private Map<UUID, Map<UUID, UUID>> templates;

    public CollectionTemplates( Map<UUID, Map<UUID, UUID>> templates ) {
        this.templates = templates;
    }

    public CollectionTemplates() {
        this( Maps.newHashMap() );
    }

    public Map<UUID, Map<UUID, UUID>> getTemplates() {
        return templates;
    }

    public synchronized void put( UUID entitySetCollectionId, UUID templateTypeId, UUID entitySetId ) {
        Map<UUID, UUID> template = templates.getOrDefault( entitySetCollectionId, Maps.newHashMap() );
        template.put( templateTypeId, entitySetId );
        templates.put( entitySetCollectionId, template );
    }

    public synchronized void putAll( Map<UUID, Map<UUID, UUID>> templateValues ) {
        for ( Map.Entry<UUID, Map<UUID, UUID>> entry : templateValues.entrySet() ) {
            Map<UUID, UUID> template = templates.getOrDefault( entry.getKey(), Maps.newHashMap() );
            template.putAll( entry.getValue() );
            templates.put( entry.getKey(), template );
        }
    }

    @Override public boolean equals( Object o ) {
        if ( this == o )
            return true;
        if ( o == null || getClass() != o.getClass() )
            return false;
        CollectionTemplates that = (CollectionTemplates) o;
        return Objects.equals( templates, that.templates );
    }

    @Override public int hashCode() {
        return Objects.hash( templates );
    }

    @Override public String toString() {
        return "CollectionTemplates{" +
                "templates=" + templates +
                '}';
    }
}

File: api/src/main/kotlin/com/openlattice/collections/CollectionTemplates.kt
package com.openlattice.collections


import com.google.common.collect.Maps
import java.util.UUID

/**
 * A Hazelcast-serializable wrapper around a MutableMap<UUID, MutableMap<UUID, UUID>>
 * for processing entity set collection templates.
 *
 * @param templates A map from entity set collection ids to their templates,
 * where their templates are maps from collection template type ids to entity set ids
 */
data class CollectionTemplates(
        val templates: MutableMap<UUID, MutableMap<UUID, UUID>> = Maps.newHashMap()
) {

    @Synchronized
    fun put(entitySetCollectionId: UUID, templateTypeId: UUID, entitySetId: UUID) {
        val template = templates.getOrDefault(entitySetCollectionId, Maps.newHashMap())
        template[templateTypeId] = entitySetId
        templates[entitySetCollectionId] = template
    }

    @Synchronized
    fun putAll(templateValues: Map<UUID, MutableMap<UUID, UUID>>) {
        for ((key, value) in templateValues) {
            val template = templates.getOrDefault(key, Maps.newHashMap())
            template.putAll(value)
            templates[key] = template
        }
    }
}